haystack/nodes/ranker/recentness_ranker.py

```diff
@@ -0,0 +1,159 @@
+                You supplied: '%s'.\n
+                Document[0] contains metadata with keys: %s.\n
+                Continuing without sorting by date.
+                """,
+                self.date_identifier,
+                ",".join(list(documents[0].meta.keys())),
+            )
+
+            return documents
+        except ParserError:
+            logger.error(
+                """
+                Could not parse date information for dates: %s\n
+                Continuing without sorting by date.
+                """,
+                " - ".join([x.meta.get(self.date_identifier, "identifier wrong") for x in documents]),
+            )
+
+            return documents
+
+        # merge scores for both documents sorted by content and date.
+        # If method is set to 'reciprocal_rank_fusion', then that is used to combine previous ranking with recency ranking.
+        # If method is set to 'score', then documents will be assigned a recency score in [0,1] and will be re-ranked based on both their recency score and their pre-existing relevance score.
+        scores_map: Dict = defaultdict(int)
+        document_map = {doc.id: doc for doc in documents}
+        weight = self.weight
+        method = self.method
+        if method not in ["reciprocal_rank_fusion", "score"]:
+            logger.error(
+                """
+                    Param <method> seems wrong.\n
+                    You supplied: '%s'.\n
+                    It should be 'reciprocal_rank_fusion' or 'score'.\n
+                    Defaulting to 'reciprocal_rank_fusion'.
+                    """,
+                method,
+            )
+            method = "reciprocal_rank_fusion"
+
+        for i, doc in enumerate(documents):
+            if method == "reciprocal_rank_fusion":
+                scores_map[doc.id] += self._calculate_rrf(rank=i) * (1 - weight)
+            elif method == "score":
+                score = float(0)
+                if doc.score is None:
+                    warnings.warn("The score was not provided; defaulting to 0")
+                elif doc.score < 0 or doc.score > 1:
+                    warnings.warn("The score is outside the [0,1] range; defaulting to 0")
+                else:
+                    score = doc.score
+
+                scores_map[doc.id] += score * (1 - weight)
+
+        for i, doc in enumerate(sorted_by_date):
+            if method == "reciprocal_rank_fusion":
+                scores_map[doc.id] += self._calculate_rrf(rank=i) * weight
+            elif method == "score":
+                scores_map[doc.id] += self._calc_recentness_score(rank=i, amount=len(sorted_by_date)) * weight
+        sorted_doc_ids = sorted(scores_map.items(), key=lambda d: d[1] if d[1] is not None else -1, reverse=True)
+
+        top_k = top_k or self.top_k or len(sorted_doc_ids)
+        docs = []
+        for idx, score in sorted_doc_ids[:top_k]:
+            doc = document_map[idx]
```

Comment:
Do we really need the map of documents here? Can't we sort the list of documents with the doc.score as key after updating the scores. Then we don't need to append one document at a time in a loop, which should be faster.
I have something like the following in mind: `return sorted(documents, key=lambda x: x.score)` after doc.score was updated for all docs.

Reply:
You can also have a look at SentenceTransformersRanker to see how it's implemented there without a need for `append`.

Reply:
I didn't find a way to fully remove documents map, but I did get rid of the docs list and appending, this should hopefully address the issue